0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+
+Similarly to set comprehensions or assign-such-that statements, the domain of a `foreach` loop
+must be enumerable. The following loop would produce the error `"the domain of a foreach loop must be enumerable,
+but Dafny's heuristics can't figure out how to produce an enumeration for 'x'"`.
+
+```dafny
+foreach x: real | 0.0 <= x < 1.0 {
+  ...
+}
+```
+
+The domain is allowed to be potentially infinite if the loop is used in a compiled context and an explicit `decreases` clause is provided.
+`decreases *` is permitted, in which the `foreach` loop may never terminate. Any other `decreases` clause can be provided
+to ensure the loop terminates even if the domain is potentially infinite. The following (slightly bizarre) example is legal:
+
+```dafny
+foreach x: nat 
+  decreases 10 - x
+{
+  print x, "\n";
+  if x == 10 { break; }
+}
+```
+
+Most `foreach` loops will take the form `foreach x | x in c { ... }`, where `c` is a `seq`, `set` or `multiset`.
+This includes expressions like `m.Keys`, `m.Values` and `m.Items` when `m` is a `map`.
+Looping over the keys and values of a map is particularly readable and aesthetically pleasing with this syntax,
+which also makes it easier for the compiled program to avoid unnecessary intermediate tuple values:
+
+```dafny
+foreach k, v | k in m.Keys && m[k] == v {
+  print "m[", k, "] == ", v;
+}
+```
+
+Note that the range expression is optional, and if omitted the loop will iterate over all
+possible values of the types of the bound variables. This is only permitted if all such types
+are enumerable, which is not true of the `real` type, for example. This supports an elegant
+pattern for iterating over simple datatypes and other types with few values, including subset types.
+
+```
+datatype Suit = Clubs | Diamonds | Hearts | Spades
+
+foreach s: Suit {
+  ...
+}
+
+foreach b1: bool, b2: bool {
+  expect TestMyFunctionWithOptions(b1, b2) == true;
+}
+
+type SmallIndex = x: nat | 0 <= x < 8
+
+foreach i: SmallIndex {
+  ...
+}
+```
+
+There is no builtin support for automatically tracking the enumerated values so far, as there is for
+`iterator` values (see https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-iterator-types). 
+It is straightforward to track these values manually, however:
+
+```dafny
+ghost var xs := [];
+foreach x | x in s {
+  ...
+  xs := xs + [x];
+}
+```
+
+It is even possible to attach this ghost state directly to the sequence with a helper function:
+
+```dafny
+foreach x, xs | (x, xs) in WithPrefixes(c) {
+  ...
+}
+```
+
+Where the signature of `WithPrefixes` is something like:
+
+```dafny
+function method WithPrefixes<T>(s: seq<T>): seq<(T, ghost seq<T>)> {
+  ...
+}
+```
+
+Similarly, a helper function can be provided to maintain a running index of the enumerated values:
+
+```dafny
+foreach x, i | (x, i) in WithIndexes(s) {
+  ...
+}
+
+// Alternatively, if s is a sequence:
+foreach x, i | 0 <= i < |s| && s[i] == x {
```

Comment:
Honestly I feel like this is more of an example of the power of using general quantifier domains to bind multiple variables, and less of an idiom I'd strongly encourage in general, since it's not much better than a `for` loop. :) It will be important for `WithIndexes` to be verification-friendly and runtime-efficient so we can move away from depending on random access.

Reply:
It's actually still better than a for loop because you can do filtering on the value itself !

Reply:
To me it's significantly better if you write it this way:

```
foreach i, x | 0 <= i < |s|
    && x == s[i]
    && x != 0 {
  ...
}
```

Indeed, everything that is an expression is almost always easier to reason about than a statement (for example, with the `for` and continue, you get the `x != 0` only implicitly after the block of the if because there is a continue, so you need to wrap your head to remember that x is not zero, whereas with the above, it's obvious)